File: problems/09_heap_priority_queue/0295_find_median_from_data_stream/solution.swift
// Problem: 0295. Find Median from Data Stream
// Difficulty: Hard

import Foundation

struct MaxHeapMF {
    var a: [Int] = []
    var count: Int { a.count }
    var top: Int { a[0] }
    mutating func push(_ x: Int) {
        // TODO: implement
    }
    @discardableResult
    mutating func pop() -> Int {
        // TODO: implement
        return 0
    }
}

struct MinHeapMF {
    var a: [Int] = []
    var count: Int { a.count }
    var top: Int { a[0] }
    mutating func push(_ x: Int) {
        // TODO: implement
    }
    @discardableResult
    mutating func pop() -> Int {
        // TODO: implement
        return 0
    }
}

class MedianFinder {
    var lo = MaxHeapMF()
    var hi = MinHeapMF()
    init() {
        // TODO: implement
    }

    func addNum(_ num: Int) {
        // TODO: implement
    }

    func findMedian() -> Double {
        // TODO: implement
        return 0
    }
}

func checkD(_ a: Double, _ e: Double, _ msg: String) {
    if abs(a - e) > 1e-9 { fatalError("FAIL \(msg): \(a) vs \(e)") }
}

// MARK: - Tests
// let s = Solution()

// Test 1
// let result = s.solve(...)

print("All tests passed!")

File: problems/09_heap_priority_queue/0295_find_median_from_data_stream/solution.py
"""
Problem: 0295. Find Median from Data Stream
Difficulty: Hard
URL: https://leetcode.com/problems/find-median-from-data-stream/

解法：max-heap + min-heap 平衡，addNum O(log n)，findMedian O(1)。
"""

from __future__ import annotations

from typing import List

import heapq

class MedianFinder:
    def __init__(self):
        pass

    def addNum(self, num: int) -> None:
        pass

    def findMedian(self) -> float:
        pass


if __name__ == "__main__":
    medianFinder = MedianFinder()

    # Test 1
    # result = medianFinder.solve(...)
    # assert result == expected

    print("All tests passed!")
